design/design.md

```diff
@@ -0,0 +1,414 @@
+3. performance: proof size, communication complexity, computational complexity.
+4. interfaces: the applications should be able to use our
+implementation relatively easily
+
+### Definitions
+
+The definitions requirement can be stated as follows.
+Taking an ALBA protocol we implement, can you construct (on paper) a larger
+protocol, such as Mithril, and mathematically prove its guarantees such as
+completeness and soundness?
+Currently, ALBA schemes satisfying the definitions in the ALBA paper cannot
+be used to construct Mithril, Peras and Leios.
+
+First, all three require a multisignature scheme secure against
+grinding attacks.
+In short, unlike the setting in the ALBA paper, the theorem statement to be
+proven is not fixed in advance, but can depend on the previous actions of
+the adversary.
+The adversary can simulate (grind) different actions to see if there is one
+that will allow him to cheat.
+For instance, the lottery in Mithril uses a seed from the Ouroboros protocol
+to pseudo-randomly select the signing committee;
+this seed can to some extent be controlled by the adversary.
+Research is needed to see how grinding can be handled and on what
+level: ALBA or the multisignature scheme.
+
+Additionally, Mithril and possibly Peras and Leios cannot provide a weight
+oracle as required by the definitions in the ALBA paper.
+The weight oracle must report the weight of any data element.
+For Mithril, if the data elements are signatures, the weight oracle would
+have to report for any signature the amount of stake of the creator of the
+signature.
+Clearly, this is impossible.
+Even if the data element was a pair of public key and signature, the weight
+oracle would have no way of verifying that the public key owns any stake.
+Again, research is needed to construct the right ALBA scheme on
+paper before implementing it.
+
+### Correctness
+
+The definitions of ALBA require, for input parameters $\lambda_{sec}$ and
+$\lambda_{rel}$, that the soundness error is at most $2^{-\lambda_{sec}}$
+and the completeness error is at most $2^{-\lambda_{rel}}$.
+Ideally, our implementation should also make this guarantee.
+
+However, at this point, this is not our goal, since achieving that would
+require significant additional effort.
+We need to roll out a reference implemenentation so that our primary clients
+(Mithril, Peras, Leios) can incorporate it in their prototypes, and also so
+that the external community can study and experiment with ALBA as well.
+
+The following is a list of known discrepancies that prevent us from making
+the rigorous guarantees about soundness and completeness errors.
+1. tiny rounding errors in parameter derivation due to the use of
+floating-point numbers
+2. tiny errors when using a uniformly random hash output to sample distributions
+needed for ALBA such as Bernoulli, uniform, binomial, Poisson
+3. lack of a rigorous analysis of the $H_1$ optimization described in
+section 3.3 of the paper
+
+### Performance
```

Comment:
Stop talking about Peras, Mithril and Alba.
This wil be talks withing IOHK with the relevant teams. It should not be in here.